plugin/trace/trace.go

```diff
@@ -120,9 +122,27 @@ func (t *trace) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 	span.SetTag(tagType, req.Type())
 	span.SetTag(tagRcode, rcode.ToString(rw.Rcode))
 
+	for key, value := range t.tags {
+		value = tagValue(ctx, value)
+		if value != "" {
+			span.SetTag(key, value)
+		}
+	}
+
 	return status, err
 }
 
+func tagValue(ctx context.Context, value string) string {
```

Comment:
I think all this can be done during setup time. In the hot path it should just call the value func. That is, change `tags` to be type `map[string]metadata.Func`. Then you just call that func that you look up (which might return a static value if there are no metadata tags in it).

I think code in https://github.com/coredns/coredns/blob/master/plugin/metadata/provider.go can handle most of this for you. To avoid depending on the metadata plugin directly, it probably makes sense to move a lot of that stuff into a plugin/pkg that is used by any plugin processing or providing metadata. But that can be a different PR.

Reply:
Ok so for the static values they'll simply be aded to the provider on setup?

That makes sense, I'll try to get this done tomorrow, thanks!

Reply:
> To avoid depending on the metadata plugin directly, it probably makes sense to move a lot of that stuff into a plugin/pkg that is used by any plugin processing or providing metadata. But that can be a different PR.

I'm struggling with this aspect of CoreDNS in general. Not all plugins seem equal. Some provide discrete functionality, others do cross cutting concerns (e.g. tracing) and cross-plugin compile time dependencies are allowed (metadata, trace, dnstap, etc). The latter is scary because it means encumbrance between plugins only increases with time. Anyway, I think you're right - further separation could be done in follow-up work.

Thanks for the quick reviews!